src/router_core/connections.c

```diff
@@ -273,8 +273,14 @@ void qdr_record_link_credit(qdr_core_t *core, qdr_link_t *link)
 void qdr_close_connection_CT(qdr_core_t *core, qdr_connection_t  *conn)
 {
     sys_mutex_lock(&conn->work_lock);
-    conn->closed = true;
-    conn->error  = qdr_error(QD_AMQP_COND_CONNECTION_FORCED, "Connection forced-closed by management request");
+    //
+    // In some cases, this function can be called on an already closed connection in which case
```

Comment:
Calling this twice seems like a bug - is this new behavior caused by this patch?  The adaptors are not calling the qdr_core_close_connection() API twice on the same qdr_connection_t, are they?  Or is the management agent attempting to re-delete a deleted connection?

Reply:
It is the latter and it happens in the http2 adaptor. When a connector is deleted in the http2 adaptor, it immediately (but asynchronously tries) to delete the associated connection (in http2 adaptor, one connector can have only one associated connection). While the core thread is actually trying to delete the connection, the management command comes in to tries to delete the connection as well. This happened 1 out of 10 times in my testing, so I added that bit code to check if the conn is already closed to prevent a qdr_error leak.

Reply:
Ah! Ok, so if I understand correctly:

1. The CI change adds a management call to delete the _connection_ - which is cool.
2. At the same moment the adaptor is servicing a previous management request to delete the _connector_, which then correctly calls qdr_core_close_connection()

So at the core we have two connection delete paths - one via the management agent and one via the action queued by the adaptor thread - both are valid and can potentially happen simultaneously "in the field".

Given that I agree the check for conn->closed is the correct move.

But - and it seems I always have a big but, right? - I wouldn't activate the connection if conn->closed was already true. That will cause the connection process to call the close handler twice which it probably shouldn't do.

See next comment.

Reply:
Used a local flag (outside of lock) to avoid activation and it worked !